libyul/backends/evm/ControlFlowGraphBuilder.cpp

```diff
@@ -124,6 +124,74 @@ void markRecursiveCalls(CFG& _cfg)
 			});
 		}
 }
+/// Marks each cut-vertex in the CFG, i.e. each block that begins a disconnected sub-graph of the CFG.
+/// Entering such a block means that control flow will never return to a previously visited block.
+void markStartsOfSubGraphs(CFG& _cfg)
+{
+	/**
+	 * Detect cut vertices following https://en.wikipedia.org/wiki/Biconnected_component#Pseudocode
+	 * Can ignore the root node, since it is never a cut vertex in our sense.
+	 * Marks terminating and exiting blocks as pseudo cut vertices (since entering such blocks will also
+	 * never return to previous blocks).
+	 */
+	static constexpr auto mark = [](CFG::BasicBlock* _root) {
+		set<CFG::BasicBlock*> visited;
+		map<CFG::BasicBlock*, size_t> depths;
+		map<CFG::BasicBlock*, size_t> low;
+		map<CFG::BasicBlock*, CFG::BasicBlock*> parent;
+		auto dfs = [&](CFG::BasicBlock* _u, size_t _depth, auto _recurse) -> void {
+			visited.insert(_u);
+			depths[_u] = low[_u] = _depth;
+
+			vector<CFG::BasicBlock*> children;
```

Comment:
```suggestion
			vector<CFG::BasicBlock*> children = _u->entries;
```

@chriseth This I think I should still change... without this, I'm visiting the graph as directed graph, but the algorithm assumes it's undirected... it does affect gas costs, sometimes a bit better, sometimes a bit worse... I'm a bit surprised that it works both ways, but also visiting the entries here is probably the safer choice - although it may be worth thinking through this a bit more at some point...

Reply:
Maybe someone should try to understand the algorithm...